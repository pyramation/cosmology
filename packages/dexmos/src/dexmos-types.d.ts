export interface DenomUnits {
  denom: string
  exponent: number
  aliases?: string[]
}

export interface Coin {
  description?: string
  denom_units: DenomUnits[]
  base: string // this is the denom
  name: string
  display: string
  symbol: string
  logo_URIs: {
    png: string
    svg?: string
  }
  coingecko_id?: string
}
