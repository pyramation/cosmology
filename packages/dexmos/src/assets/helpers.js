// need to figure out typescript compatability
// import { Coin } from '../dexmos-types';
import assetList from './assetlist.json';

// /**
//  * 
//  * @param {string} denom 
//  * @returns {Coin}
//  */
export const getCoinFromDenom = (denom) => {
    for (let asset of assetList.assets) {
        if (asset.base === denom) {
            return asset;
        }
    }
}