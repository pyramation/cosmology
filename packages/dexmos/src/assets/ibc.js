// https://github.com/osmosis-labs/osmosis-frontend/blob/master/src/config.ts#L531
export const ibcChannels = [
	{
		counterpartyChainId: 'cosmoshub-4',
		sourceChannelId: 'channel-0',
		destChannelId: 'channel-141',
		coinMinimalDenom: 'uatom',
	},
	{
		counterpartyChainId: 'columbus-5',
		sourceChannelId: 'channel-72',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'uluna',
	},
	{
		counterpartyChainId: 'crypto-org-chain-mainnet-1',
		sourceChannelId: 'channel-5',
		destChannelId: 'channel-10',
		coinMinimalDenom: 'basecro',
	},
	{
		counterpartyChainId: 'columbus-5',
		sourceChannelId: 'channel-72',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'uusd',
	},
	{
		counterpartyChainId: 'secret-4',
		sourceChannelId: 'channel-88',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'uscrt',
	},
	{
		counterpartyChainId: 'juno-1',
		sourceChannelId: 'channel-42',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ujuno',
	},
	{
		counterpartyChainId: 'stargaze-1',
		sourceChannelId: 'channel-75',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ustars',
	},
	{
		counterpartyChainId: 'chihuahua-1',
		sourceChannelId: 'channel-113',
		destChannelId: 'channel-7',
		coinMinimalDenom: 'uhuahua',
	},
	{
		counterpartyChainId: 'core-1',
		sourceChannelId: 'channel-4',
		destChannelId: 'channel-6',
		coinMinimalDenom: 'uxprt',
	},
	{
		counterpartyChainId: 'columbus-5',
		sourceChannelId: 'channel-72',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'ukrw',
	},
	{
		counterpartyChainId: 'akashnet-2',
		sourceChannelId: 'channel-1',
		destChannelId: 'channel-9',
		coinMinimalDenom: 'uakt',
	},
	{
		counterpartyChainId: 'regen-1',
		sourceChannelId: 'channel-8',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'uregen',
	},
	{
		counterpartyChainId: 'sentinelhub-2',
		sourceChannelId: 'channel-2',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'udvpn',
	},
	{
		counterpartyChainId: 'irishub-1',
		sourceChannelId: 'channel-6',
		destChannelId: 'channel-3',
		coinMinimalDenom: 'uiris',
	},
	{
		counterpartyChainId: 'iov-mainnet-ibc',
		sourceChannelId: 'channel-15',
		destChannelId: 'channel-2',
		coinMinimalDenom: 'uiov',
	},
	{
		counterpartyChainId: 'emoney-3',
		sourceChannelId: 'channel-37',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ungm',
	},
	{
		counterpartyChainId: 'emoney-3',
		sourceChannelId: 'channel-37',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'eeur',
	},
	{
		counterpartyChainId: 'microtick-1',
		sourceChannelId: 'channel-39',
		destChannelId: 'channel-16',
		coinMinimalDenom: 'utick',
	},
	{
		counterpartyChainId: 'likecoin-mainnet-2',
		sourceChannelId: 'channel-53',
		destChannelId: 'channel-3',
		coinMinimalDenom: 'nanolike',
	},
	{
		counterpartyChainId: 'impacthub-3',
		sourceChannelId: 'channel-38',
		destChannelId: 'channel-4',
		coinMinimalDenom: 'uixo',
	},
	{
		counterpartyChainId: 'bitcanna-1',
		sourceChannelId: 'channel-51',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'ubcna',
	},
	{
		counterpartyChainId: 'bitsong-2b',
		sourceChannelId: 'channel-73',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ubtsg',
	},
	{
		counterpartyChainId: 'kichain-2',
		sourceChannelId: 'channel-77',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'uxki',
	},
	{
		counterpartyChainId: 'panacea-3',
		sourceChannelId: 'channel-82',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'umed',
	},
	{
		counterpartyChainId: 'bostrom',
		sourceChannelId: 'channel-95',
		destChannelId: 'channel-2',
		coinMinimalDenom: 'boot',
	},
	{
		counterpartyChainId: 'comdex-1',
		sourceChannelId: 'channel-87',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'ucmdx',
	},
	{
		counterpartyChainId: 'cheqd-mainnet-1',
		sourceChannelId: 'channel-108',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ncheq',
	},
	{
		counterpartyChainId: 'lum-network-1',
		sourceChannelId: 'channel-115',
		destChannelId: 'channel-3',
		coinMinimalDenom: 'ulum',
	},
	{
		counterpartyChainId: 'vidulum-1',
		sourceChannelId: 'channel-124',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'uvdl',
	},
	{
		counterpartyChainId: 'desmos-mainnet',
		sourceChannelId: 'channel-135',
		destChannelId: 'channel-2',
		coinMinimalDenom: 'udsm',
	},
	{
		counterpartyChainId: 'dig-1',
		sourceChannelId: 'channel-128',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'udig',
	},
];