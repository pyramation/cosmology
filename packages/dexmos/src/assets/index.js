import json from './assetlist.json';
export const assets = json.assets;
// https://raw.githubusercontent.com/osmosis-labs/assetlists/main/osmosis-1/osmosis-1.assetlist.json
