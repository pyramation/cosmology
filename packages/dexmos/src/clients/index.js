export * from './osmosis';
export * from './cosmos';