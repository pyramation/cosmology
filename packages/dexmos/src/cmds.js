
import _create_mnemonic_wallet from './commands/create-mnemonic-wallet';
import _delegate from './commands/delegate';
import _get_balances from './commands/get-balances';
import _get_latest_block from './commands/get-latest-block';
import _get_pool_apr from './commands/get-pool-apr';
import _get_price from './commands/get-price';
import _get_staking_rewards from './commands/get-staking-rewards';
import _get_transaction from './commands/get-transaction';
import _ibc_transfer from './commands/ibc-transfer';
import _ibc_wip from './commands/ibc-wip';
import _join_pool from './commands/join-pool';
import _list_apis from './commands/list-apis';
import _list_pools from './commands/list-pools';
import _list_prices from './commands/list-prices';
import _load_recipe from './commands/load-recipe';
import _lock_tokens from './commands/lock-tokens';
import _send from './commands/send';
import _swap_exact_amount_in from './commands/swap-exact-amount-in';
import _wallet_from_mnemonic from './commands/wallet-from-mnemonic';
const Commands = {};
Commands['create-mnemonic-wallet'] = _create_mnemonic_wallet;
Commands['delegate'] = _delegate;
Commands['get-balances'] = _get_balances;
Commands['get-latest-block'] = _get_latest_block;
Commands['get-pool-apr'] = _get_pool_apr;
Commands['get-price'] = _get_price;
Commands['get-staking-rewards'] = _get_staking_rewards;
Commands['get-transaction'] = _get_transaction;
Commands['ibc-transfer'] = _ibc_transfer;
Commands['ibc-wip'] = _ibc_wip;
Commands['join-pool'] = _join_pool;
Commands['list-apis'] = _list_apis;
Commands['list-pools'] = _list_pools;
Commands['list-prices'] = _list_prices;
Commands['load-recipe'] = _load_recipe;
Commands['lock-tokens'] = _lock_tokens;
Commands['send'] = _send;
Commands['swap-exact-amount-in'] = _swap_exact_amount_in;
Commands['wallet-from-mnemonic'] = _wallet_from_mnemonic;

  export { Commands }; 

  