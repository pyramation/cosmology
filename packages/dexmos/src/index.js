export * from './assets';
export * from './clients';
export * from './utils/osmo';
export * from './driver/driver';
export * from './messages';