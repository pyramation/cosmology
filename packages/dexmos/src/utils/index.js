export * from './chain';
export * from './prompt';
export * from './helpers';
export * from './wallet';
export * from './prompts/osmo';
