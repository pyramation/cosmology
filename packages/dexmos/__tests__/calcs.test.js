// @ts-nocheck
import pricesFixture from '../__fixtures__/coingecko/api/v3/simple/price/data.json';
import bankFixture from '../__fixtures__/keplr/bank/balances/osmo1x/data.json';
import poolsFixture from '../__fixtures__/keplr/osmosis/gamm/v1beta1/pools/data.json';
import lockedPoolsFixture from '../__fixtures__/keplr/osmosis/lockup/v1beta1/account_locked_coins/osmo1/data.json';
import { Token } from '../src/model/Token';
import { assets, chains } from '@pyramation/cosmos-registry';
import { assets as osmosisAssets } from '../src/assets';
import cases from 'jest-in-case';
import { displayUnitsToDenomUnits, baseUnitsToDisplayUnits, getOsmosisSymbolIbcName } from '../src/utils';
import Long from 'long';
import {
    getCoinGeckoIdForSymbol,
    getSymbolForCoinGeckoId,
    osmoDenomToSymbol,
    symbolToOsmoDenom,
    OsmosisToken,
    convertCoinToDisplayValues,
    convertCoinsToDisplayValues,
    calculateCoinsBalance,
    convertPricesToDenomPriceHash,
    getPoolByGammName,
    getUserPools,
    convertPoolToDisplayValues,
    convertPoolsToDisplayValues,
    getFilteredPoolsWithValues,
    symbolsAndDisplayValuesToCoinsArray,
    getTradesRequiredToGetBalances
} from '../src/utils/osmo';

cases('getCoinGeckoIdForSymbol', opts => {
    expect(getCoinGeckoIdForSymbol(opts.name)).toBe(opts.geckoId);
}, [
    { name: 'ATOM', geckoId: 'cosmos' },
    { name: 'AKT', geckoId: 'akash-network' },
    { name: 'OSMO', geckoId: 'osmosis' }
]);

cases('getSymbolForCoinGeckoId', opts => {
    expect(getSymbolForCoinGeckoId(opts.geckoId)).toBe(opts.name);
}, [
    { name: 'ATOM', geckoId: 'cosmos' },
    { name: 'AKT', geckoId: 'akash-network' },
    { name: 'OSMO', geckoId: 'osmosis' }
]);

cases('displayUnitsToDenomUnits', opts => {
    expect(displayUnitsToDenomUnits(opts.name, opts.amount)).toBe(opts.value);
}, [
    { name: 'ATOM', amount: 10, value: 10000000 },
    { name: 'AKT', amount: .6, value: 600000 },
    { name: 'OSMO', amount: 10, value: 10000000 }
]);

cases('osmoDenomToSymbol', opts => {
    expect(osmoDenomToSymbol(opts.denom)).toBe(opts.name);
}, [
    { name: 'ATOM', denom: 'ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2' },
    { name: 'AKT', denom: 'ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4' },
    { name: 'OSMO', denom: 'uosmo' }
]);

cases('symbolToOsmoDenom', opts => {
    expect(symbolToOsmoDenom(opts.name)).toBe(opts.denom);
}, [
    { name: 'ATOM', denom: 'ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2' },
    { name: 'AKT', denom: 'ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4' },
    { name: 'OSMO', denom: 'uosmo' }
]);

cases('convertPricesToDenomPriceHash', opts => {
    expect(convertPricesToDenomPriceHash(opts.prices)).toEqual(opts.hash);
}, [
    {
        name: 'test1',
        prices: { 'akash-network': { usd: 100 }, cosmos: { usd: 20 } },
        hash: {
            "ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4": 100,
            "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2": 20
        }
    }
]);

cases('getPoolByGammName', opts => {
    const prices = convertPricesToDenomPriceHash(pricesFixture)
    const pools = getFilteredPoolsWithValues({ prices, pools: poolsFixture.pools })
    expect(getPoolByGammName(pools, opts.name).id).toEqual(opts.poolId);
}, [
    {
        name: 'gamm/pool/1',
        poolId: "1"
    },
    {
        name: 'gamm/pool/600',
        poolId: "600"
    }
]);

describe('basic portfolio', () => {
    const coins = symbolsAndDisplayValuesToCoinsArray(
        [
            {
                symbol: 'ATOM',
                amount: 10
            },
            {
                symbol: 'AKT',
                amount: 1000
            },
            {
                symbol: 'OSMO',
                amount: 100
            }
        ]
    );

    const prices = convertPricesToDenomPriceHash(pricesFixture)

    it('calculate portfolio value (balances)', () => {
        const value = calculateCoinsBalance({ prices, coins });
        expect(value).toBe(3086.3);
    });

    it('calculate pool value (global)', () => {
        const values = getFilteredPoolsWithValues({ prices, pools: poolsFixture.pools })
        const [first, second] = values;
        expect(first).toMatchSnapshot();
        expect(second).toMatchSnapshot();
    });

    it('calculate pool value (user LP)', () => {
        const pools = getFilteredPoolsWithValues({ prices, pools: poolsFixture.pools })
        const result = getUserPools({ pools, lockedPools: lockedPoolsFixture.coins })
        expect(result).toMatchSnapshot();
    });
});

describe('user actions', () => {
    const balances = symbolsAndDisplayValuesToCoinsArray(
        [
            {
                symbol: 'ATOM',
                amount: 10
            },
            {
                symbol: 'AKT',
                amount: 1000
            },
            {
                symbol: 'OSMO',
                amount: 100
            }
        ]
    );

    const prices = convertPricesToDenomPriceHash(pricesFixture)

    it('cannot desire more value than current', async () => {
        const desired = symbolsAndDisplayValuesToCoinsArray(
            [
                {
                    symbol: 'ATOM',
                    amount: 100000
                }
            ]
        );
        let failed = false;
        try {
            getTradesRequiredToGetBalances({ prices, balances, desired })
        } catch (e) {
            failed = true;
        }
        expect(failed).toBe(true);
    });

    it('no trades required', async () => {
        const desired = symbolsAndDisplayValuesToCoinsArray(
            [
                {
                    symbol: 'ATOM',
                    amount: 10
                },
                {
                    symbol: 'OSMO',
                    amount: 10
                }
            ]
        );
        const trades = getTradesRequiredToGetBalances({ prices, balances, desired })
        expect(trades.length).toBe(0);
    });

    it('trades required', async () => {
        const desired = symbolsAndDisplayValuesToCoinsArray(
            [
                {
                    symbol: 'AKT',
                    amount: 1050
                },
                {
                    symbol: 'OSMO',
                    amount: 100
                }
            ]
        );
        const trades = getTradesRequiredToGetBalances({ prices, balances, desired })
        console.log(JSON.stringify(trades, null, 2));
    });
});


it('calcs', async () => {
    // console.log(prices);
    // console.log(bank);
    // console.log(pools);
    // let driver = new Driver()
    // const jobs = await driver.getAllJobs([
    //   { "type": "coin", "coin": "UST", "weight": 0.3 },
    //   { "type": "pool", "pool": { "id": 562, "coin1": "LUNA", "coin2": "UST", "balance": 0.5 }, "weight": 0.3 },
    //   { "type": "pool", "pool": { "id": 611, "coin1": "ATOM", "coin2": "STARS", "balance": 0.7 }, "weight": 0.4 }
    // ])
    // console.log(jobs);
    // expect(jobs.length).toEqual(10)

    // expect(jobs[0].job.inputCoin).toEqual("LUNA")
    // expect(jobs[0].job.targetCoin).toEqual("UST")
    // expect(jobs[0].job.amount).toBeGreaterThan(0)

    // expect(jobs[5].job.inputCoin).toEqual("UST")
    // expect(jobs[5].job.targetCoin).toEqual("STARS")
    // expect(jobs[5].job.amount).toBeGreaterThan(0)
})
