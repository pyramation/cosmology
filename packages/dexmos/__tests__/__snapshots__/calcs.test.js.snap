// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic portfolio calculate pool value (global) 1`] = `
Object {
  "@type": "/osmosis.gamm.v1beta1.Pool",
  "address": "osmo1mw0ac6rwlp5r8wapwk3zs6g29h8fcscxqakdzw9emkne6c8wjp9q0t3v8t",
  "displayPoolAssets": Array [
    Object {
      "allocation": 0.5,
      "symbol": "ATOM",
      "token": Object {
        "amount": "6196127370623",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
      },
      "value": Object {
        "amount": "6196127370623",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
        "displayAmount": 6196127.370623,
        "value": 226964145.5859205,
      },
      "weight": "536870912000000",
    },
    Object {
      "allocation": 0.5,
      "symbol": "OSMO",
      "token": Object {
        "amount": "24164665575940",
        "denom": "uosmo",
      },
      "value": Object {
        "amount": "24164665575940",
        "denom": "uosmo",
        "displayAmount": 24164665.57594,
        "value": 219898456.741054,
      },
      "weight": "536870912000000",
    },
  ],
  "future_pool_governor": "24h",
  "id": "1",
  "poolAssets": Array [
    Object {
      "token": Object {
        "amount": "6196127370623",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
      },
      "weight": "536870912000000",
    },
    Object {
      "token": Object {
        "amount": "24164665575940",
        "denom": "uosmo",
      },
      "weight": "536870912000000",
    },
  ],
  "poolParams": Object {
    "exitFee": "0.000000000000000000",
    "smoothWeightChangeParams": null,
    "swapFee": "0.003000000000000000",
  },
  "pricePerShare": 1.353191082134662e-18,
  "pricePerShareDenom": 1.353191082134662,
  "totalShares": Object {
    "amount": "330228752041469100486167962",
    "denom": "gamm/pool/1",
  },
  "totalValue": 446862602.3269745,
  "totalWeight": "1073741824000000",
}
`;

exports[`basic portfolio calculate pool value (global) 2`] = `
Object {
  "@type": "/osmosis.gamm.v1beta1.Pool",
  "address": "osmo1500hy75krs9e8t50aav6fahk8sxhajn9ctp40qwvvn8tcprkk6wszun4a5",
  "displayPoolAssets": Array [
    Object {
      "allocation": 0.8,
      "symbol": "ION",
      "token": Object {
        "amount": "1289632142",
        "denom": "uion",
      },
      "value": Object {
        "amount": "1289632142",
        "denom": "uion",
        "displayAmount": 1289.632142,
        "value": 15871722.009058138,
      },
      "weight": "858993459200000",
    },
    Object {
      "allocation": 0.2,
      "symbol": "OSMO",
      "token": Object {
        "amount": "374852643996",
        "denom": "uosmo",
      },
      "value": Object {
        "amount": "374852643996",
        "denom": "uosmo",
        "displayAmount": 374852.643996,
        "value": 3411159.0603636,
      },
      "weight": "214748364800000",
    },
  ],
  "future_pool_governor": "24h",
  "id": "2",
  "poolAssets": Array [
    Object {
      "token": Object {
        "amount": "1289632142",
        "denom": "uion",
      },
      "weight": "858993459200000",
    },
    Object {
      "token": Object {
        "amount": "374852643996",
        "denom": "uosmo",
      },
      "weight": "214748364800000",
    },
  ],
  "poolParams": Object {
    "exitFee": "0.000000000000000000",
    "smoothWeightChangeParams": null,
    "swapFee": "0.005000000000000000",
  },
  "pricePerShare": 3.335651016095471e-20,
  "pricePerShareDenom": 0.03335651016095471,
  "totalShares": Object {
    "amount": "578084487147375945198234662",
    "denom": "gamm/pool/2",
  },
  "totalValue": 19282881.06942174,
  "totalWeight": "1073741824000000",
}
`;
